### ISSUE TITLE
How i can return data from script?

### ISSUE THREAD

**Issue by Sanceilaks** [nushell/nushell#13372]
### Question

For example i have a script
```nu
def main [] {
   { a: 123 }
}
```
but describe says that it is byte stream
```sh
❯ nu example.nu | describe
byte stream
```

how i can return data from script?

**Comment by WindSoilder** [comment]
When you run `nu example.nu` directly, it will render the data and return it.  Not sure what's the real issue?

**Comment by suimong** [comment]
I guess OP's intention might be that `nu example.nu | describe` results in `record<a: int>`, to make further piping possible?

Currently (v0.95.0), `nu` seems to collect and deserialize the output of a script into string, which can be demonstrated by the following:

```nushell
# example.nu
def main [
  --str
  --int
  --record
] {
   if $str {return "abc"} else if $int {return 123} else if $record {return {a: "efg"}} else {}
}
```

```shell
/tmp/tmp.dfSV0i2AHb
❯ nu example.nu --str | str upcase
ABC

/tmp/tmp.dfSV0i2AHb
❯ nu example.nu --record | str upcase
╭───┬─────╮
│ A │ EFG │
╰───┴─────╯

/tmp/tmp.dfSV0i2AHb
❯ nu example.nu --int | str reverse
321

/tmp/tmp.dfSV0i2AHb
❯ nu example.nu --record | collect | describe
string
```

However, with `nu example.nu --record`, Nushell correctly renders the return value as a record by syntax highlighting the key, indicating that it's possible for Nushell to know the data type of a bytestream. 

AFAIK there isn't a command that performs such "deserialization", but it would be very useful.

**Comment by Sanceilaks** [comment]
> I guess OP's intention might be that `nu example.nu | describe` results in `record<a: int>`, to make further piping possible?
> 
> Currently (v0.95.0), `nu` seems to collect and deserialize the output of a script into string, which can be demonstrated by the following:
> 
> ```nushell
> # example.nu
> def main [
>   --str
>   --int
>   --record
> ] {
>    if $str {return "abc"} else if $int {return 123} else if $record {return {a: "efg"}} else {}
> }
> ```
> 
> ```shell
> /tmp/tmp.dfSV0i2AHb
> ❯ nu example.nu --str | str upcase
> ABC
> 
> /tmp/tmp.dfSV0i2AHb
> ❯ nu example.nu --record | str upcase
> ╭───┬─────╮
> │ A │ EFG │
> ╰───┴─────╯
> 
> /tmp/tmp.dfSV0i2AHb
> ❯ nu example.nu --int | str reverse
> 321
> 
> /tmp/tmp.dfSV0i2AHb
> ❯ nu example.nu --record | collect | describe
> string
> ```
> 
> However, with `nu example.nu --record`, Nushell correctly renders the return value as a record by syntax highlighting the key, indicating that it's possible for Nushell to know the data type of a bytestream.
> 
> AFAIK there isn't a command that performs such "deserialization", but it would be very useful.

Yeah, that's exactly what I meant. I would like to be able to get the record from the script

**Comment by bminer** [comment]
> When you run `nu example.nu` directly, it will render the data and return it. Not sure what's the real issue?

The issue is that the nushell process calling the nushell script is unaware of the structured data; it only sees the bytestream, much like calling any other external process.

Put another way, when you type `nu example.nu` into nushell A, it spawns a nushell process B. Process B understands the structured data and prints it to the bytestream. It would be nice if process B somehow "knew" that it was being invoked by nushell (process A) and would print the serialized data to the bytestream instead (i.e. maybe encoded as a nuon). Then nushell running in process A could decode the bytestream and reconstruct the structured data.

**Comment by GrabbenD** [comment]
> nushell process calling the nushell script is unaware of the structured data; it only sees the bytestream, much like calling any other external process.

For future readers who stumble upon issues such as `nu::shell::cannot_pass_list_to_external`, this and related tickets _(like https://github.com/nushell/nushell/issues/10371, https://github.com/nushell/nushell/issues/3551, https://github.com/nushell/nushell/issues/15176, https://github.com/nushell/nushell/issues/9373, https://github.com/nushell/nushell/issues/13372)_ are being tracked here:

**https://github.com/nushell/nushell/issues/15219**

---

**Workarounds** (please let me know if there's more!):
- https://github.com/nushell/nushell/issues/10371#issue-1897002593
  Export the function and import it: `$ use ./foo.nu; foo { echo "foo" }`
- https://github.com/nushell/nushell/issues/15176#issuecomment-2683725788
  Import the argument as a string and serialize it into Nuon
- https://github.com/nushell/nushell/issues/9373#issuecomment-1635919687
  Another example of converting the payload into Nuon but with pipes
- https://github.com/nushell/nushell/issues/13372#issuecomment-2447280655
  Transfer data using environment variables

**Comment by 132ikl** [comment]
Hi, I'm going to close this as a duplicate of #9373 for the sake of keeping track of things in one place. If there's some aspect of this issue you think isn't tracked by #9373 then let me know and I can add details there, or if necessary, reopen this issue
